src/CircuitZKit.ts

```diff
@@ -0,0 +1,224 @@
+import fs from "fs";
+
+import path from "path";
+
+import * as snarkjs from "snarkjs";
+import * as https from "https";
+import { Calldata, CircuitZKitConfig, CircuitZKitPrivateConfig, CompileOptions, Inputs, ProofStruct } from "./types";
+import { createDirIfNotExists, ensureFileExists } from "./utils";
+
+const { CircomRunner, bindings, Context } = require("@distributedlab/circom2");
+
+export class CircuitZKit<I extends Inputs> {
+  public readonly config: CircuitZKitPrivateConfig;
+
+  constructor(config: CircuitZKitConfig) {
+    const circuitId = path.parse(config.circuitFile).name;
+    const verifierId = `${circuitId}Verifier`;
+
+    this.config = {
+      circuitFile: config.circuitFile,
+      artifactsDir: config.artifactsDir,
+      circuitOutDir: config.circuitOutDir,
+      verifierOutDir: config.verifierOutDir,
+      r1csFile: path.join(config.circuitOutDir, `${circuitId}.r1cs`),
+      zKeyFile: path.join(config.circuitOutDir, `${circuitId}.zkey`),
+      vKeyFile: path.join(config.circuitOutDir, `${circuitId}.vkey.json`),
+      wasmFile: path.join(config.circuitOutDir, `${circuitId}_js`, `${circuitId}.wasm`),
+      verifierFile: path.join(config.verifierOutDir, `${verifierId}.sol`),
+      ptauDir: path.join(config.artifactsDir, ".ptau"),
+      ptauFile: "",
+      circuitId,
+      verifierId,
+      wasmBytes: fs.readFileSync(require.resolve("@distributedlab/circom2/circom.wasm")),
+      groth16Template: fs.readFileSync(path.join(__dirname, "templates", "verifier_groth16.sol.ejs"), "utf8"),
+    };
+  }
+
+  public async compile(options = {} as CompileOptions): Promise<void> {
+    createDirIfNotExists(this.config.circuitOutDir);
+
+    const args = this._getCompileArgs({
+      sym: options.sym ?? true,
+      c: options.c ?? false,
+    });
+
+    await this._compile(args, options.quiet ?? true);
+
+    await this._fetchPtauFile();
+    await this._generateZKey();
+    await this._generateVKey();
+  }
+
+  public async generateProof(inputs: I): Promise<ProofStruct> {
+    ensureFileExists(this.config.zKeyFile);
+    ensureFileExists(this.config.wasmFile);
+
+    return (await snarkjs.groth16.fullProve(inputs, this.config.wasmFile, this.config.zKeyFile)) as ProofStruct;
+  }
+
+  public async verifyProof(proof: ProofStruct): Promise<boolean> {
+    ensureFileExists(this.config.vKeyFile);
+
+    const verifier = JSON.parse(fs.readFileSync(this.config.vKeyFile).toString());
+
+    return await snarkjs.groth16.verify(verifier, proof.publicSignals, proof.proof);
+  }
+
+  public async createVerifier(): Promise<void> {
```

Comment:
Since we already have a verification key, should we render the verifier manually to have more control over it

cc https://github.com/iden3/snarkjs/blob/master/src/zkey_export_solidityverifier.js